### PR TITLE
Sign in

### PR DESCRIPTION
**Pull request by blhohmann**
# Pull Request: Sign-In Feature

## Description

This pull request adds a sign-in feature to the project. Users can now log in to access their accounts.

## Changes Made

Login Changes:
- Added new sign-in functionality.
- Add view for /login with a barebones login form
- Added routes for /login and /logout
- Added controller for logging in and out, session_controller
- Added testing for session_controller
- Added testing for the session requests HTTP responses
- Flash messages for correct and incorrect logins as well as logging out

General Changes:
- Added coverage folder to .gitignore
- Added a user factory for ease of testing
- Pulled an early version of the create-user branch for testing purposes
- Updated seeding of users to use password instead of password_digest
- Add flash notice and flash alert to application.html.erb so all pages will show flash messages, also some styling added for those flash messages
- Added a dummy home page and routing to test my flash messages after redirect

### To manually test the sign-in feature:
Happy Path:
1. Visit '/login' in your local development environment.
2. Enter valid user credentials after seeding the database.
3. Click the "Log In" button.
4. Verify that you are redirected to '/'.
Sad Path:
1.  Visit '/login' in your local development environment.
2. Enter invalid user credentials after seeding the database.
3. Click the "Log In" button.
4. Verify that you are not redirected to '/' and instead stay on the page.

#### To run the rspec tests put this in your terminal
``` bundle exec rspec spec/controllers/sessions_controller_spec.rb```

## Impact on Existing Code
- No breaking changes are introduced.
- User seeding needed to change

## Screenshots or Demos
Sad Path login:
image
image

Happy Path login:
image
image

Sad Path logout:
image
image

Happy Path logout:
image
image

## Additional Notes

- This feature is part of the ongoing effort to improve user experience and security.
- No major performance issues have been observed during testing.
- If you seeded the database with the previous seed that went straight to password_digest and try to login as admin@admin.com; admin you will encounter a bug because it was improperly seeded. To fix, on my branch run
```
bin/rails db:drop
bin/rails db:create
bin/rails db:migrate
bin/rails db:seed
```

## Reviewer Requests

I kindly request a review of these changes. Please provide feedback and suggestions. As this was a 2 point user story just 1 review should suffice.

## Status Checks and Tests

- All automated tests have passed.
- Code coverage is ```Coverage report generated for RSpec to /mnt/c/Users/blake/selt2023-final-project-team-001/coverage. 167 / 168 LOC (99.4%) covered.```
- No known issues or limitations at this time.

## Discussion

Feel free to comment, ask questions, or provide feedback.

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -7,3 +7,28 @@
...
+
+users = [
+  {
+    first_name: "admin",
+    last_name: "admin",
+    email: "admin@admin.com",
+    password: "admin",
+    password_confirmation: "admin",
+    is_seller: false,
+    is_buyer: false
```
Comment (evfinkn) [inline db/seeds.rb:19]:
```suggestion
    is_buyer: false,
    is_admin: true
```

File: db/seeds.rb [hunk 2]
```diff
@@ -7,3 +7,28 @@
...
+  },
+  {
+    first_name: "admin2",
+    last_name: "admin2",
+    email: "admin2@admin.com",
+    password: "admin2",
+    password_confirmation: "admin2",
+    is_seller: false,
+    is_buyer: false
```
Comment (evfinkn) [inline db/seeds.rb:28]:
```suggestion
    is_buyer: false,
    is_admin: true
```

File: app/views/sessions/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+  <div>
+    <%= form.label :password %>
+    <%= form.password_field :password %>
+  </div>
+
+  <div>
+    <%= form.submit 'Log In' %>
+  </div>
+<% end %>
```
Comment (evfinkn) [inline app/views/sessions/new.html.erb:15]: This should be updated to use `bootstrap_form_with` like `users/new.html.erb` does. Not positive this works, but basing it off of that file:

```suggestion
<%= bootstrap_form_with model: @user, url: login_submit_path, method: :post, local: true do |f| %>
  <%= f.email_field :email, label: "Email: " %>
  <%= f.password_field :password, label: "Password: " %>
  <%= f.submit "Log In" %>
<% end %>
```

**Review by evfinkn — CHANGES_REQUESTED** [review]

File: config/database.yml
[the remark this answers is not in the document]
Comment (blhohmann) [reply config/database.yml:1]: It is now reset to version in main

File: .gitignore [hunk 4]
```diff
@@ -36,3 +36,4 @@
 
 # Jetbrains IDEs
 /.idea
+/coverage/
```
[the remark this answers is not in the document]
Comment (blhohmann) [reply .gitignore]: Okay, I added it again, and removed from gitignore.

File: db/seeds.rb (hunk 2 above)
Comment (blhohmann) [reply db/seeds.rb]: Changed admin and admin2 to have is_admin be true.

File: app/views/sessions/new.html.erb (hunk 3 above)
Comment (blhohmann) [reply app/views/sessions/new.html.erb]: I changed it as you suggested, and it still passes rspec tests and it worked with my manual testing.

**Review by evfinkn — APPROVED** [review]
Look's good to merge 🚀

**Pull request merged**